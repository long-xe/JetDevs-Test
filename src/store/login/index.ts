import { reducer } from "./login.slice";

export default {
    reducer,
}