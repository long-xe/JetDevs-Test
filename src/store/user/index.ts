import { reducer } from "./user.slice";
import * as actions from './user.action'

export default {
    reducer,
    ...actions,
}