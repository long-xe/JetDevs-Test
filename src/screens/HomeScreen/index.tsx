import { View, FlatList } from 'react-native';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getUserFromApiAction, refreshUserFromApiAction } from '../../store/user/user.action';
import type { AppDispatch, RootState } from '../../store/store';
import { styles } from './style';
import UserCard from '../../components/cards/home/userCard';

const results = 10;

export default function HomeScreen() {
  const dispatch = useDispatch<AppDispatch>();
  const { users, favoriteUser } = useSelector(
    (state: RootState) => state.users,
  );

  const [page, setPage] = useState(0);
  const [isRefresh, setIsRefresh] = useState(false);

  useEffect(() => {
    dispatch(getUserFromApiAction({ page: page, results }));
  }, []);


  const onLoadMore = () => {
    dispatch(getUserFromApiAction({ page: page + 1, results }));
    setPage(e => e + 1);
  };

  const onRefresh = () => {
    setIsRefresh(true);
    dispatch(refreshUserFromApiAction({page: 0, results}));
    setIsRefresh(false);
  };

  return (
    <View style={styles.container}>
      <FlatList
        initialNumToRender={10}
        maxToRenderPerBatch={1}
        data={users}
        keyExtractor={(item, idx) => `${idx + item.login.uuid}`}
        refreshing={isRefresh}
        onRefresh={onRefresh}
        onEndReached={({ distanceFromEnd }) => {
          if (distanceFromEnd < 0) return;
          onLoadMore();
        }}
        renderItem={({ item, index }) => {
          return <UserCard user={item} index={index} isFavorite={favoriteUser.some(e => e.login.uuid === item.login.uuid)}/>;
        }}
      />
    </View>
  );
}
