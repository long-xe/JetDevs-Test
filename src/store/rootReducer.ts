
import login from './login';
import user from './user';

export const rootReducer = {
  users: user.reducer,
  login: login.reducer
};
