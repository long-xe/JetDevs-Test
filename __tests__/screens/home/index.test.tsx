import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { act } from 'react-test-renderer';
import { store } from '../../../src/store/store';
import HomeScreen from '../../../src/screens/HomeScreen';


describe('home screen', () => {
    it('should renders screen correctly', () => {
      const page = render(
          <Provider store={store}>
            <HomeScreen />
          </Provider>
      );
      expect(page).toMatchSnapshot();
    });
  });