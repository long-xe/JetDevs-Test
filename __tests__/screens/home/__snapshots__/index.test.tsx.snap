// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home screen should renders screen correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "marginHorizontal": 20,
      "marginVertical": 10,
    }
  }
>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
